lib/actionpack/all/actionpack.rbi

```diff
@@ -43,12 +43,57 @@ class ActionController::Parameters
   sig { returns(T::Boolean) }
   def permitted?; end
 
-  sig { params(key: T.any(Symbol, T::Array[Symbol])).returns(ActionController::Parameters) }
+  sig do
+    params(
+      key: T.any(String, Symbol),
+    ).returns(
+      T.nilable(
+        T.any(
+          String,
+          Numeric,
+          ActionController::Parameters,
+        ),
+      ),
+    )
+  end
+  def [](key); end
+
+  sig do
+    params(
+      key: T.any(
+        String,
+        Symbol,
+        T::Array[T.any(String, Symbol)],
+      ),
+    ).returns(
+      T.any(
+        String,
+        Numeric,
+        T::Array[T.untyped],
+        ActionController::Parameters,
+      ),
+    )
+  end
   def require(key); end
 
   # required is an alias of require
-  sig { params(key: T.any(Symbol, T::Array[Symbol])).returns(ActionController::Parameters) }
-  def required(key); end
+  sig do
+    params(
+      key: T.any(
+        String,
+        Symbol,
+        T::Array[T.any(String, Symbol)],
+      ),
+    ).returns(
+      T.any(
+        String,
+        Numeric,
```

Comment:
Does it actually return Strings/Numbers? This makes it much harder to write code like so:

```ruby
def book_params
  params.require(:book).permit(:author_id, :title)
end
```

Since `permit` won't exist on the other types this can return.

Reply:
> Does it actually return Strings/Numbers?

yep

```ruby
pry(main)> ActionController::Parameters.new({foo: 1}).require(:foo)
=> 1

pry(main)> ActionController::Parameters.new({foo: 'foo'}).require(:foo)
=> "foo"

pry(main)> ActionController::Parameters.new({foo: {bar: 'baz'}}).require(:foo)
=> <ActionController::Parameters {"bar"=>"baz"} permitted: false>

pry(main)> ActionController::Parameters.new({foo: 1, bar: 1}).require([:foo, :bar])
=> [1, 1]
```

> This makes it much harder to write code like so:

oh totally agreed. i wrote these helpers a week ago which have helped with the types: https://gist.github.com/frewsxcv/fc5d7426921bb64f5a6821721cff4581

Reply:
that's... really weird.

How does `params.require(:foo).permit(:bar)` in controllers work, then? Is `params` actually something other than an instance of `ActionController::Parameters`?

Reply:
Oh, it seems to looks like this:

```ruby
ActionController::Parameters.new({foo: { bar: 'test'}}).require(:foo).permit(:bar)
=> <ActionController::Parameters {"bar"=>"test"} permitted: true>
```

Reply:
yeah, this is a gnarly api. i don't really understand how to write fullproof code with it, without a bunch of `is_a?` checks...

Reply:
also, adding onto this, since `params` can [be populated from a json request](https://guides.rubyonrails.org/action_controller_overview.html#json-parameters), any scalar json datatype can be returned from `require`, including a number